pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.io.util;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.pinot.common.utils.StringUtil;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+/**
+ * Implementation of {@link ValueReader} that will allow each value to be of variable
+ * length and there by avoiding the unnecessary padding.
+ *
+ * The layout of the file is as follows:
+ * <p> Header Section: </p>
+ * <ul>
+ *   <li> Magic header byte sequence. </li>
```

Comment:
Adding the following fields in the header will help with compatibility issues:

1. Version number
2. Start offset of (non-header) data. This way there is no implicit assumption on size of header and that data starts where header ends.

Reply:
I've added version number in the most recent commit.
About 2, having such offset will force us to read that value before reading any `byte[]` from the store. Currently, the read side code is simple because you know where the data starts already and can directly read. So, my take is, unless there is a strong reason, it might be better off leaving it like this.

Reply:
For 2, one benefit is that if later we want to add more metadata in the header without changing the rest of the format, we can do so without having to create a new version number of worrying about backward compatibility. The old reader will still be able to read the new file (with new metadata), it will just not see the new metadata.

Reply:
Fixed this and pushed my changes again.